### PR TITLE
Clear Cache in Data Dictionary

### PR DESCRIPTION
**Pull request by zandre-eng**
## Technical Summary

Link to ticket [here](https://dimagi.atlassian.net/browse/SC-3331).

This PR reduces the cache time for `get_used_props_by_case_type()` to ten minutes. When data is added to a domain (e.g. form submissions or case imports) a case property will be incorrectly marked as safe to delete until the cache for the util function is cleared. Decreasing the cache time will help to reduce the lag between when data is added to a domain and when a case type/prop is correctly flagged as unsafe to delete.

Additionally, this PR also clears the cache for `get_data_dict_case_types()` on case type/prop save or delete. There are various places on HQ, such as the case type filter in reports, that showed a case type even after it has been deleted. This is because the results were being cached. Clearing the cache on case type/prop save or delete helps to ensure that these deleted items are not shown around HQ.

## Feature Flag

None

### Safety story

- Local testing
- Only afffects caching

### Automated test coverage

N/A

### QA Plan

No QA planned.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by mkangia** [comment]
Looking good. Holding off on review since this is in draft state.

**Review by kaapstorm — APPROVED** [review]

File: corehq/apps/data_dictionary/models.py
Comment (mkangia) [inline corehq/apps/data_dictionary/models.py:40]: Do we need to also expire `get_data_dict_props_by_case_type`?

**Comment by mkangia** [comment]
Just FYI that I branched of your work to create this PR https://github.com/dimagi/commcare-hq/pull/34154 that addresses some more caching concerns

**Comment by zandre-eng** [comment]
Taking this out of draft, so @mkangia this is ready for a final review.

Just also noting that this PR will only get merged once the sub-branch has been merged into this one.

File: corehq/apps/data_dictionary/util.py
Comment (mkangia) [inline corehq/apps/data_dictionary/util.py:385]: nit: is there a way to asses the impact this would have on DB with cache expiring earlier?
Its only used when loading DD right? Which I believe should be okay since that page is only available for advance plan and higher.

**Review by mkangia — APPROVED** [review]

**Comment by mkangia** [comment]
> Just also noting that this PR will only get merged once the sub-branch has been merged into this one.

Hey @zandre-eng 
Is this for the branch I mentioned for https://github.com/dimagi/commcare-hq/pull/34154?
You don't need to block this PR for it.

**Comment by zandre-eng** [comment]
> Hey @zandre-eng
Is this for the branch I mentioned for https://github.com/dimagi/commcare-hq/pull/34154?
You don't need to block this PR for it.

@mkangia I thought it might be easier to merge into this branch and then have everything merged to master. I suppose it's not completely necessary. Thanks for letting me know, I will then go ahead and merge in this branch.

File: corehq/apps/data_dictionary/util.py
Comment (zandre-eng) [reply corehq/apps/data_dictionary/util.py:385]: The only real way I can think of for assessing this would be to do a performance test both with and without the delete feature enabled. It's worth noting however, that QA has done a performance test on this and concluded that it didn't have a noticable impact on performance.

Given the above, I think it should be fine with the reduced cache expiry.

Comment (mkangia) [reply corehq/apps/data_dictionary/util.py:385]: Thanks @zandre-eng 
Where can I see details of the performance test done? I was requesting something similar lately so would be helpful to see how QA does this.

Comment (zandre-eng) [reply corehq/apps/data_dictionary/util.py:385]: You can have a look at this QA ticket for more context around the [performance testing](https://dimagi.atlassian.net/browse/QA-5997) that was done.

**Comment by kaapstorm** [comment]
@mkangia :
> Just FYI that I branched of your work to create this PR #34154 that addresses some more caching concerns

I really quite like this workflow -- branching off a PR to add or amend things. I was thinking about it over the weekend:
* Whether we should do this more
* When we should
* When we shouldn't

I'll continue this conversation in Slack.

**Pull request merged**
